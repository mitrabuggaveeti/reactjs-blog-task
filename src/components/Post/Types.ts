import { InewState as Props } from '../Feed/Types';

export interface Iprops {
    post : {
        userId: number
        id: number
        title: string
        body: string
    },
    posts : Props['post'],
    setPosts : React.Dispatch<React.SetStateAction<Props['post']>>
}

export interface InewState {
    comment : {
        data : {
            postId : number,
            id : number,
            email : string,
            body : string
        }[],
        error : string
    }
}