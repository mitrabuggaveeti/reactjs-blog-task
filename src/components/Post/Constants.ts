export const COMMENTS_LABEL_SHOW = 'Show Comments'
export const COMMENTS_LABEL_HIDE = 'Hide Comments'
export const DELETE_POPUP_TITLE = 'Confirm to Delete'
export const DELETE_POPUP_MESSAGE = 'Are you sure to do this delete this post :'
export const CONFIRM_DELETE = 'YES'
export const CANCEL_DELETE = 'NO'
export const LABEL_DELETE = 'Delete'
export const LABEL_EDIT = 'Edit'
export const DATA_LOAD_LABEL = 'Loading'