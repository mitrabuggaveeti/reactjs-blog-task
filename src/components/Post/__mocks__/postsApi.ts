export const fakeData = {
   data :[
    {
        "postId": 1,
        "id": 1,
        "name": "id labore ex et",
        "email": "Eliseo@gardner.biz",
        "body": "laudantium enim quasi est"
      },
      {
        "postId": 1,
        "id": 2,
        "name": "quo vero rei",
        "email": "Jayne_Kuhic@sydney.com",
        "body": "est natus enim nihil"
      },
      {
        "postId": 1,
        "id": 3,
        "name": "quo vero rei",
        "email": "Jayne_Kuhic@sydney.com",
        "body": "est natus enim nihil"
      }
    ]
}

export const getAllComments = async (id:number) =>{

    return await new Promise(resolve =>{
        resolve(fakeData)
    })
}

export  const deletePost = async (id:number) =>{
  
    return await new Promise(resolve =>{
      resolve({ ok : true})
    })
}


