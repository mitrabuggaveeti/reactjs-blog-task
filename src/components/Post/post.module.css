.card {
    box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
    transition: 0.3s;
    margin-left: 2%;
    margin-right: 2%;
    margin-top: 1%;
}

.card:hover {
    box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
}

.container {
    padding: 2px 16px;
}

.title{
    font-size: larger;
    font-family: revert;
    margin: 6px 0 6px 0;
}

.description{
    margin: 12px 0 12px 0;
}

.commentsBtn{
    position: absolute;
    right: 0;
}
.commentsBtn , .editBtn, .deleteBtn {
    background-color: #5eb461;
    border: none;
    color: white;
    padding: 5px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 13px;
    margin: 4px 2px;
    cursor: pointer;
    border-radius: 4px;
}

.editBtn{
    background-color: rgb(40, 172, 165);
}

.deleteBtn{
    background-color: rgb(241, 109, 109);
}


.btns{
    margin: 6px 0 6px 0;
    position: relative;
}

