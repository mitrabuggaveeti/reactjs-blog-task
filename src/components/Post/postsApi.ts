export const getAllComments = async (id:number) =>{

    try{
        const response = await fetch(`https://jsonplaceholder.typicode.com/posts/${id}/comments`);
        const data = await response.json();

        if(response.ok){
            return {
                data,
                error :''
            }
        }
        return {
                data : [],
                error :'Failed in fetching the comments'
        }
    }
    catch(error){
        return {
            data : [],
            error
        }
    }
}

export  const deletePost = async (id:number) =>{

    try {
        const response = await fetch(`https://jsonplaceholder.typicode.com/posts/${id}`, {
            method: 'DELETE',
         });

         if(response.ok){
             return {
                 status : 'deleted',
                 error : ''
             }
         }
         else return {
             status : 'pending',
             error : 'failed'
         }
    
        
    } catch (error) {
        return {
            error,
            status: 'failed'
        }
    }
}


