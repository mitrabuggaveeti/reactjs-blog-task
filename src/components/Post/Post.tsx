import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import Comment from '../Comments/Comment';
import {deletePost, getAllComments} from './postsApi';
import styles from './post.module.css';
import * as postConstants from './Constants';
import { confirmAlert } from 'react-confirm-alert'; 
import 'react-confirm-alert/src/react-confirm-alert.css'; 
import { InewState, Iprops } from './Types';

const Post : React.FC<Iprops> = (props) => {

    const {title,body,id, userId} = props.post;
    const {posts,setPosts} = props;
    let history = useHistory();

    const [commentsLabel,setCommentsLabel] = useState(false);
    const [comments,setComments] = useState<InewState['comment']>({
      error : postConstants.DATA_LOAD_LABEL,data:[]
    });
    
   const handleShowComments  = async() =>{
      if(!commentsLabel){
        const data = await getAllComments(id);
        setComments(data);
      }
      setCommentsLabel((prevState) => !prevState);
    }


    const handleDeleteConfirmaton = async () =>{
       const data =  await deletePost(id); 
        if(!data.error){
            const postsAfterDeletion = posts.data.filter((post) => post.id !== id);
            setPosts({data :postsAfterDeletion,error:''});
        } 
    }

    const handleDeletePost = async() =>{

        confirmAlert({
            title: postConstants.DELETE_POPUP_TITLE,
            message: `${postConstants.DELETE_POPUP_MESSAGE} ${id}`,
            buttons: [
              {
                label: postConstants.CONFIRM_DELETE,
                onClick: () => handleDeleteConfirmaton()
              },
              {
                label: postConstants.CANCEL_DELETE,
                onClick: () => {}
              }
            ]
          });   
    }

    const handleEditPost = ()=>{
        history.push(`/edit/${id}/${userId}`);
    }

    const commentsError = () => {
       return !comments.error  ? <Comment comments={comments.data.slice(0,10)}/> : <h2>{comments.error}.....</h2>
    }

    return (
        <div className={styles.card}>
            <div className={styles.container}>
                <div className={styles.title}>
                   <strong> {title} </strong>
                </div>
                <div className={styles.description}>
                   {body}
                </div>
                <div className={styles.btns}>                  
                     <button className={styles.editBtn} onClick={handleEditPost}>{postConstants.LABEL_EDIT}</button>
                     <button className={styles.deleteBtn} onClick={handleDeletePost} data-testid='deleteComments'>{postConstants.LABEL_DELETE}</button>
                     <button className={styles.commentsBtn} onClick={handleShowComments} data-testid='showComments'>
                      {
                         !commentsLabel ? postConstants.COMMENTS_LABEL_SHOW : postConstants.COMMENTS_LABEL_HIDE
                      }
                     </button>
                </div>
                      {
                        commentsLabel ?  commentsError() :''
                      }
            </div>
        </div>
    )
}

export default Post
