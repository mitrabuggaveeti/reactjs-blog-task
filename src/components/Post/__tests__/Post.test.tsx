import { cleanup, fireEvent, render,waitFor} from "@testing-library/react";
import Post from "../Post";
import * as services from '../postsApi';


jest.mock('../postsApi');

const MOCK_DATA = {
    post : 
      {
        "userId": 1,
        "id": 1,
        "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
        "body": "quia et suscipit"
      },
    posts :{
        data :[
            {
                "userId": 2,
                "id": 2,
                "title": "optio reprehenderit",
                "body": "quia et suscipit"
            },
            {
                "userId": 3,
                "id": 3,
                "title": "sunt aut facere ",
                "body": "est autem sunt rem"
            }
        ],
        error : ''
    }
    
}

describe('Post component', () => {
    afterEach(cleanup);

    it('Should have comments rendered on button click',async () => {
                
        const { getByTestId, asFragment } = render(<Post post={MOCK_DATA.post} posts={MOCK_DATA.posts} setPosts={()=>{}}/>);

        fireEvent.click(getByTestId('showComments'));

        const listNode = await waitFor(() => getByTestId('allComments'));

        expect(listNode.children).toHaveLength(3);
        expect(asFragment()).toMatchSnapshot();
    });

    it('Should change comments button label', async ()=>{
        const { getByTestId} = render(<Post post={MOCK_DATA.post} posts={MOCK_DATA.posts} setPosts={()=>{}}/>);

        expect(getByTestId('showComments')).toHaveTextContent('Show Comments');
        fireEvent.click(getByTestId('showComments'));

        const commentsBtn = await waitFor(() => getByTestId('showComments'));
        expect(commentsBtn.innerHTML).toEqual('Hide Comments');

    })


    it('Should delete blog with id:1', async ()=>{

        const mockAddListener = jest.spyOn(services, 'deletePost');

        const { getByTestId} = render(<Post post={MOCK_DATA.post} posts={MOCK_DATA.posts} setPosts={()=>{}}/>);

        fireEvent.click(getByTestId('deleteComments'));

        expect(mockAddListener).toHaveBeenCalledWith(1);
    })

})
