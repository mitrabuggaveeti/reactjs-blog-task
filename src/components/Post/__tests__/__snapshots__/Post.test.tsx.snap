// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component Should have comments rendered on button click 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="container"
    >
      <div
        class="title"
      >
        <strong>
           sunt aut facere repellat provident occaecati excepturi optio reprehenderit 
        </strong>
      </div>
      <div
        class="description"
      >
        quia et suscipit
      </div>
      <div
        class="btns"
      >
        <button
          class="editBtn"
        >
          Edit
        </button>
        <button
          class="deleteBtn"
          data-testid="deleteComments"
        >
          Delete
        </button>
        <button
          class="commentsBtn"
          data-testid="showComments"
        >
          Hide Comments
        </button>
      </div>
      <div>
        <h2>
          Comments
        </h2>
        <ul
          data-testid="allComments"
        >
          <li>
            laudantium enim quasi est
            <p>
               - Eliseo@gardner.biz
            </p>
          </li>
          <li>
            est natus enim nihil
            <p>
               - Jayne_Kuhic@sydney.com
            </p>
          </li>
          <li>
            est natus enim nihil
            <p>
               - Jayne_Kuhic@sydney.com
            </p>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
